### DISCUSSION TITLE
Ignore some dependencies based on PHP version

### DISCUSSION THREAD

**Discussion by andreamk** [composer/composer#10947] [Q&A / Support] [answered]
Hello I have been searching for a long time for a solution but can't find it. 

I have created a PHP library that works on PHP 5.4+
https://github.com/andreamk/JsonSerialize
the composer
https://github.com/andreamk/JsonSerialize/blob/20220612_AL_JS_php_stan/composer.json

The library runs unit tests on PHP 5.4, 7, 8 and 8.1 on every commit via github workflows.
This works fine, depending on the PHP version the correct version of unit tests is installed .

The problem arose when I added PHPStan to perform static analysis.
PHPstan does not have a working version on PHP 5.4.
This in theory would not be a problem because static code analysis is not performed on PHP 5.4

I created 3 different workflows
https://github.com/andreamk/JsonSerialize/tree/20220612_AL_JS_php_stan/.github/workflows

- One where code analysis is performed (phpsniff, phpstan) with the default PHP version
- One where unit tests are run on PHP 7 and 8
- And one where unit tests are run on php 5.4

the last workflow fails because the composer gives error when installing dependencies because of phpstan which would not be needed in this specific workflow.
https://github.com/andreamk/JsonSerialize/runs/7370003797?check_suite_focus=true

Any idea how I could configure the composer to install only unit tests and skip phpstan on php 5.4 ?

I realize this is an intricate problem, I hope someone has the solution.

thanks
Andrea

**Comment by Seldaek** [comment, marked as answer]
We have a similar problem on Composer 2.2 branch which supports PHP 5.3+, what we do there is PHPStan is conditionally required for the PHPStan builds https://github.com/composer/composer/blob/2.2/.github/workflows/phpstan.yml#L55-L57 - but you could do the opposite, and make sure you add a `composer remove phpstan/phpstan` step to your PHP 5.4 test build.

**Comment by andreamk** [comment]
@Seldaek Thanks for reply, I figured out how to do it.

**Comment by andreamk** [comment]
Hi @Seldaek 

https://github.com/andreamk/JsonSerialize/blob/main/.github/workflows/phpunit_php5.yml

I finally found a solution; your suggestion helped me realize that I had the wrong focus. 
It is not possible to solve the problem via composer config but it becomes trivial working at docker level.

Very simply for unit test workflow on PHP 5.4/5.6 before installing dependecies I remove references to PHPstan in composer.json.
```
    - name: Remove phpstan from composer
         run: sed -i "/\bphpstan\b/d" composer.json
```

This way I can work around the problem without any further changes.

Thank you very much
Andrea
